src/datumaro/util/multi_procs_util.py

```diff
@@ -0,0 +1,91 @@
+# Copyright (C) 2023 Intel Corporation
+#
+# SPDX-License-Identifier: MIT
+
+import logging as log
+from contextlib import contextmanager
+from enum import IntEnum
+from queue import Full, Queue
+from threading import Condition, Thread
+from typing import Any, Iterator, TypeVar
+
+__all__ = ["consumer_generator"]
+
+
+class ProducerMessage(IntEnum):
+    START = 0
+    END = 1
+
+
+Item = TypeVar("Item")
+
+
+@contextmanager
+def consumer_generator(
+    producer_generator: Iterator[Item],
+    queue_size: int = 100,
+    enqueue_timeout: float = 5.0,
+    join_timeout: float = 10.0,
+) -> Iterator[Item]:
+    """Context manager that creates a generator to consume items produced by another generator.
+
+    This context manager sets up a producer thread that generates items from the `producer_generator`
+    and enqueues them to be consumed by the consumer generator, which is also created by this function.
+
+    Parameters:
+        producer_generator: A generator that produces items.
+        queue_size: The maximum size of the shared queue between the producer and consumer.
+        enqueue_timeout: The maximum time to wait for enqueuing an item to the queue if it's full.
+        join_timeout: The maximum time to wait for the producer thread to finish when exiting the context.
+
+    Returns:
+        Iterator: A context for iterating over the generated items.
+    """
+    queue = Queue(maxsize=queue_size)
+    lock = Condition()
+    is_terminated = False
+
+    def _enqueue(item: Any, queue: Queue):
+        while True:
+            try:
+                queue.put(item, block=True, timeout=enqueue_timeout)
+                return
+            except Full:
+                with lock:
+                    if is_terminated:
+                        raise RuntimeError(
+                            "Item to enqueue is left. However, the main process is terminated."
+                        )
+
+    def _target(queue: Queue) -> None:
+        try:
+            _enqueue(ProducerMessage.START, queue)
+
+            for item in producer_generator():
```

Comment:
It seems `producer_generator` is `Iterator`, so they could be use directly by `pruducer_generator`.
```suggestion
            for item in producer_generator:
```

Reply:
True, https://github.com/openvinotoolkit/datumaro/pull/1145/commits/865fb471695a9635ddd37b76383e1adc746aac1c fixes some wrong typings about it.